feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/Opaque.java

```diff
@@ -0,0 +1,41 @@
+package dev.diplomattest.somelib;
+
+import dev.diplomattest.somelib.ntv.somelib_h;
+import dev.diplomattest.somelib.ntv.somelib_h;
+
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.ref.Cleaner;
+
+public class Opaque {
+    private MemorySegment internal;
+    private final Cleaner.Cleanable cleanable;
+    static class OpaqueCleaner implements Runnable {
+        MemorySegment segment;
+        OpaqueCleaner(MemorySegment segment) {
+            this.segment = segment;
+        }
+        public void run() {
+            somelib_h.Opaque_destroy(this.segment);
+        }
+    }
+    public Opaque() {
+        var invoker = somelib_h.Opaque_new.makeInvoker();
+        var segment = invoker.apply();
+        this.internal = segment;
+        var opaqueCleaner = new OpaqueCleaner(segment);
+        this.cleanable = Main.cleaner.register(this, opaqueCleaner);
+    }
+
+    public long pointer() {
+        return internal.address();
+    }
+
+    public void delete() {
+        somelib_h.Opaque_destroy(this.internal);
+    }
+
+    public void assertStruct(MyStruct struct) {
```

Comment:
question: wait, jextract generated this from the C header files? How did it know that these C header files correspond to a class with methods?

Reply:
Sadly no, that would be amazing 😃. That’s just a hand rolled wrapper for a more ergonomic type. I’m using this to build the templating for opaque types

Reply:
Everything in the `ntv` package was generated using jextract from the c backend.

Reply:
Aha! Okay, carry on!